### ISSUE TITLE
Store activity identifier in database to include the activity identifier with old organisation identifier

### ISSUE THREAD

**Issue by sarinasindurakar** [younginnovations/iatipublisher#1343]
# Overview 
Currently, the activity identifier is not stored in the database. Consequently, when a user changes the organization identifier, the identifiers of existing activities undergo modifications. To address this issue, it is essential to store the IATI identifier text in the database. Likewise, both the old and new organization identifiers must be stored in the database. So, if an organization modifies its identifier, existing published activities should not be updated with the new organization identifier; they should retain their old identifiers.

# QA Spec
- [ ] Published activities must preserve their old identifier and should not change.
- [ ] The existing activities that are on the draft will also change and have the identifier according to the new organization identifier.
- [ ]  Following changes to the organization identifier, new activities should be generated with the updated organization identifier in both manual addition and imports.
- [ ] UI should accurately reflect the recent changes in the organization identifier.

**Comment by PG-Momik** [comment]
Hello @odscrachel @siwhitehouse 

>[!Note]
>When reading this, please note that : Organization 'element-name' refers to element-name in organization level which can be accessed via the Organization Data page. Likewise Activity 'element-name' refers to element-name in activity level, accessed via Activity detail page

**Requirement Summary**
Here's the summary of the requirement (as per my understanding):
1. An organization is allowed to change `organisation-identifier`
2. Change in `organisation-identifier` in Organization data will not affect published activities. 
3. The activity `iati-identifier` is not affected by change in organisation `organisation-identifier` for published activities.
4. The status of published activities will not change from 'published' to 'draft/republish' in case organization-identifier is changed.

**Things to note**
- In the current validation, organization `reporting-org >> ref` needs to be same as organization `organisation-identifier`.
- So there's a mechanism that updates organization `reporting-org` when organization `organisation-identifier` is updated.
- On January of 2023 there was a new requirement, [Rules for reporting-org](https://github.com/younginnovations/iatipublisher/pulse) which required activity `reporting-org` to  be derived from organization `reporting-org`
and changes made to reporting-org in organization level  will instantly be reflected in all activities (published/ unpublished/ re-publishable).

**Problem**
Changing Organization `orgnaization-identifier` will update organization `reporting-org`, this change will be reflected in activity `reporting-org` , subsequently changing the status from Published to Draft. This goes against requirement no.4 

**Conclusion**
After an internal discussion. Here's what we will be implementing.

| Change at organization level | Other affected element | Published Activities | Unpublished Activity | 
| ---- | ---- | ---- | ---- |
| organisation-identifier | organization `reporting-org` | Will not  affect activity `iati-identifier`  nor `reporting-org`. | Will update activity `iati-identifier` and `reporting-org`. | 
| reporting-org |  | Will not  affect activity `reporting-org`. | Will update activity `reporting-org`. |  

---

Please let us know if there's any confusion or changes to be made.

**Comment by sarinasindurakar** [comment]
- [ ]  The changes are not reflected in draft activities when  user changes the organization identifier.

[Untitled_ Jan 19,2024 10_28 AM.webm](https://github.com/younginnovations/iatipublisher/assets/65848732/c7a87d3b-5be0-4029-b348-42515233593a)

**Comment by praweshsth** [comment]
@odscrachel As discussed yesterday on the changes in reporting-org of a published activity, please confirm if we should make change or not.

**Comment by PG-Momik** [comment]
@odscrachel Could you tell us how you'd like activities to be handled in these scenarios:

> [!IMPORTANT]
> ## Context
> 1. Lets assume that an organization published an activity on 2023/12/31. 
> 2. Some time later the organization changes its `organisation-identifier` say 2024/01/02.
> 3. The activity iati-identifier does not change.
> 4. The user then proceeds to change a value in an activity (say `activity-status >> code` ).
> 5.  Because of change made in point 4, the activity status changes from 'Published' to 'Draft'.
> 6. Now the user has the option to 'Unpublish' and 'Republish'.

### Scenario 1 - Unpublish
Lets say the user unpublishes the activity, should the activity iati-identifier of this activity still use the previous organization-identifier . (organization-identifier prior to 2024/01/02 i.e the activity iaiti-identifier of this activity on 2023/12/31) or should the activity iati-identifier of this activity now use the new organization-identifier.  (organization-identifier updated on 2024/01/02)

### Scenario 2 - Republish
Lets say the user republishes the activity, should the activity be published with the activity iati-identifier it currently has (current activity iati-identifier uses organization-identifier prior to 2024/02/02) or should the activity be published with activity iati-identifier that uses the newer organization-identifier.

---

> [!Note]
> ## Things to note
> - Changes made in #1341 , will not let users to edit activity identifier once its published (both in published state or republish state).
> - Once activity is in draft state, whether it be Republish or Unpublish or newly created activity, it will be affected if organization-identifier is changed.

**Comment by praweshsth** [comment]
@IsabelBirds @odscrachel 

Re: 'It doesn't matter if an activity is updated, unpublished, republished, etc - if it has ever been published at all, the iati-identifier must persist.' :

The iati-identifier can be persisted in all cases except for unpublished. Once the activity is unpublished, the activity goes to draft stage and as of now, the activity identifier changes as per the new org-id in the system. To persist it though, we have to keep track of the org-id and provide users to select the activity identifier based on different org-id. But we have to look into the validation of activity when publishing if the validation looks for matching org-id and org-id part of the activity. 

Re: 'If an org-id is changes, the following areas should be updated':

Reporting org and participating org will change but the user will have to republish the activity again to reflect the changes in IATI Registry.

Re: other identifier, type B1 (Previous Reporting Organisation Identifier:

User will have to add it manually.

**Comment by praweshsth** [comment]
@IsabelBirds @odscrachel @dan-odsc 

Regarding the persistence of the activity identifier when user un-publishes the activity, we have a solution:

It starts from the creation of an activity. Currently, activity identifier is created using the existing Org-ID and the user is not able to change the activity identifier. User can't change it when user is editing the activity identifier. The solution is that the system will create the activity identifier when a new activity is added but the user will be able to change the activity identifier. Hence, user can change the org-id part in the activity identifier as per old org-id. That will solve the issue to persist activity identifier even when user un-publishes the activity and publish it again. 

In summary:

1. User will be able to change the activity identifier when the user creates the activity
2. User will be able to change the activity identifier when the user edits an activity
3. When org-id is changed to new one, the reporting org in all activities will be changed to new org-id. Hence all the published activities will be in republish state. The user will have to republish those activities
4. In new activities, the user will have to add old org-id in other identifier

Do you think, we understood it well and the solution will work? Please give us the feedback on this and we will move ahead accordingly.

**Comment by robredpath** [comment]
Thanks for this, @praweshsth . 

I think we're mixing up two different problems here:
* Once an activity has been published, its activity identifier must never change - even if the activity is later unpublished, edited, and republished. 
* When an organisation changes their org ID, their new org ID should be used for any new activities that are created, and it would be helpful if IATI Publisher could update their existing data with appropriate references to their new org ID. 

This issue is about the first of those cases; if we want to consider the second then we can discuss that on our next call. However it's not a very common event, I think, so it may not be a priority. 

Obviously, I don't know how your system works, but here's a couple of ways that might help to think about it:
* When the user clicks the "publish" button for the first time, the "activity ID" field is permanently set to read-only, so that no further changes can be made either by the user or the system. 
* Or, an internal boolean field is added to the database, called `has_ever_been_published`. This is `false` to start with, and then is set to `true` when the user clicks "publish" for the first time. It can never be set back to `false`. The effect of this field is that both the user interface and any system processes can no longer make changes to the activity ID. 

I hope that's useful clarification and simplification of this issue: what's important here is that we ensure that activity IDs never change, regardless of what else we change about the data!

**Comment by praweshsth** [comment]
@robredpath Thank you. We have mixed those up. For now, we will complete this issue and we can discuss and create new issue for org-id change and it's effect.

**Comment by praweshsth** [comment]
cc: @dan-odsc 

Please review it in staging.

**Comment by emmajclegg** [comment]
Thanks @praweshsth - I'll review tomorrow in staging.

**Comment by emmajclegg** [comment]
Hi @robredpath , @praweshsth - I know we've tried to separate out problem 1) "make sure activity IDs don't ever change" from problem 2) "dealing with the case where an org changes its org-ID".

It looks like we've already dealt with problem 1) from the user-perspective in this previous closed issue ([#1341](https://github.com/younginnovations/iatipublisher/issues/1341)), so I've nothing more to add from looking at the user interface alone. I can see a user can't edit the activity ID once an activity has been published, but can while it remains unpublished, which is correct.

So @praweshsth - on that basis, I'm happy to accept and close this issue in order to focus on problem 2) in the new issue you've created (#1364). I'll edit the description of #1364 to try and make the problem and expected tool behaviour as clear as possible.
